### PR TITLE
[MMS] Scaling Speech Technology to 1,000+ Languages | Add attention adapter to Wav2Vec2

### LINKED ISSUE #23811: Metas MMS speech recognition
### Feature request

Our request a simpler and more convenient inference process for a speech recognition model based on MMS just like wav2vec 2.0 in Transformers. 

### Motivation

We aim to encapsulate the various subroutines called by Facebook’s official model into a direct speech recognition model that is as easy to use as other transformer-based models like wav2vec 2.0. But we also know that the Hugging face team has been among the industry leaders in this area of work.

### Your contribution

We recognize that it may not be feasible for us to directly assist the Hugging Face technical team in this task. We believe that such an effort would be forward-looking given the popularity of MMS in current speech recognition research. The resulting model would be ideal for quickly transcribing our meeting notes.

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR adds the MMS models fine-tuned on speech recognition. 
See official announcement here: https://about.fb.com/news/2023/05/ai-massively-multilingual-speech-technology/
See more details here: https://github.com/facebookresearch/fairseq/blob/main/examples/mms/README.md#asr

Fixes #23811 and #23665

For now checkpoints are uploaded here:

## Pretrained-only
- https://huggingface.co/patrickvonplaten/mms-300m
- https://huggingface.co/patrickvonplaten/mms-1b

## ASR fine-tuned
- https://huggingface.co/patrickvonplaten/mms-1b-fl102
- https://huggingface.co/patrickvonplaten/mms-1b-l1107
- https://huggingface.co/patrickvonplaten/mms-1b-all

The fine-tuned checkpoints are based on **Adapter** layers as can be seen in this PR. The ASR fine-tuned weights consist of two parts:
- The non-adapter weights which are exactly the same as the base model weights
- Language specific fine-tuned adapter layer weights. This means we have 1000+ adapter weights for `mms-1b-all`

If one wants to use a specific language, specific adapter weights need to be loaded into `mms-1b-all`. 
By default `mms-1b-all` et. al load the English adapter layer weights as is currently done in https://huggingface.co/patrickvonplaten/mms-1b-all

The following works with this PR:
```py
from transformers import Wav2Vec2ForCTC, AutoProcessor
import soundfile as sf
import torch

ckpt = "./mms-1b-fl102/"
ckpt = "./mms-1b-l1107"
ckpt = "./mms-1b-all/"

processor = AutoProcessor.from_pretrained(ckpt)
model = Wav2Vec2ForCTC.from_pretrained(ckpt)

# get audio.flac from https://huggingface.co/datasets/patrickvonplaten/audios/blob/main/audio.flac
audio, sr = sf.read("./audio.flac")

inputs = processor(audio, sampling_rate=sr, return_tensors="pt")

with torch.no_grad():
    logits = model(**inputs).logits

transcription = processor.batch_decode(logits.argmax(-1))[0]

print(f"Transcription: {transcription}")
```

Now, the question what API to we want to build for allow the user to easily switch between languages for the fine-tuned weights.

**Note**:
- To switch from one language to another, both the tokenizer's vocab and the model's adapter layers need to be switched out
- The tokenizer can always easily hold all langs dicts in RAM because each lang has around 150 entries so we have 150,000 entries which is not too much for RAM
- **However**, things are a bit more tricky for the model. The base model requires 3.1 GB in FP32 RAM and each adapter weights are around 9MB in size. This means loading all adapter layers into RAM would cost ~9GB which is quite a bit. 

How should we design this model? We need to have some kind of switching between languages function anyways. I see the following APIs that could work.

### 1.) By default, we download **all** adapter layers and load all in RAM, but we provide a functionality to remove all language but one from RAM:

```py
from transformers import Wav2Vec2ForCTC, AutoProcessor

ckpt = "./mms-1b-all/"

processor = AutoProcessor.from_pretrained(ckpt)
model = Wav2Vec2ForCTC.from_pretrained(ckpt)  # requires at least 10GB of CPU RAM

target_lang = "esp"

processor.set_lang("esp")
adapter_id = processor.lang_to_id["esp"]
model.set_adapter_weights(adapter_id) # throw away all but one weights => 3.1GB of CPU RAM

model.to("cuda")
```

A problem with this is though also that it's not trivial to switch between languages because one needs to load the whole model again and then set the language again. Also we would have to add a `set_adapter_weights` function to Wav2Vec2 which is not ideal

### 2.) By default we only the adapter weights one of language (e.g. English) and the load upon request more adapter layers

```py
```py
from transformers import Wav2Vec2ForCTC, AutoProcessor

ckpt = "./mms-1b-all/"

processor = AutoProcessor.from_pretrained(ckpt)
model = Wav2Vec2ForCTC.from_pretrained(ckpt)  # requires only 3GB of CPU RAM

target_lang = "esp"

processor.set_lang("esp")
model.load_adapter("esp") # This will load a file called "adapter.esp.bin" from: https://huggingface.co/patrickvonplaten/mms-1b-all , cache it and replace the adapter

model.to("cuda")
```

Think this is quite user-friendly, intuitive and this way we also never require more than 3.1 GB of RAM. It however requires to add a pretty specific `load_adapter` function to Wav2Vec2 (think it's fine though).

### 3.) We just upload 1000+ repos one for each language. This way we don't need any "set" or "load" function and we just tread each adapter weights as their own model:

```py
from transformers import Wav2Vec2ForCTC, AutoProcessor

ckpt = "./mms-1b-all-esp/" # repo names then become lang specific

processor = AutoProcessor.from_pretrained(ckpt)
model = Wav2Vec2ForCTC.from_pretrained(ckpt)  # requires only 3GB of CPU RAM
model.to("cuda")
```

Big disadvantage is that it's pretty wasteful since an adapter layer is just 0.3% of all the models weights.

=> Overall, I'm tending to API **2.)** because it's the most user-friendly and intuitive. It'd just require to add a somewhat specific "load_adapter" function to Wav2Vec2, but think that's totally fine.

Thoughts @sanchit-gandhi @Vaibhavs10 @sgugger @LysandreJik @amyeroberts ?

### CODE REVIEW

**Comment by sgugger** [comment]
I think 2 is probably the better solution, and I would also make it possible to set the lang in the `from_pretrained` call:
```py
from transformers import Wav2Vec2ForCTC, AutoProcessor

ckpt = "./mms-1b-all/"

processor = AutoProcessor.from_pretrained(ckpt)
model = Wav2Vec2ForCTC.from_pretrained(ckpt, target_lang="esp")

processor.set_lang("esp")

model.to("cuda")

### Stuff
# want to change the language:
model.load_adapter("fra") 
```

**Comment by amyeroberts** [comment]
Agreed, with all above - 2 would be my choice:

* 1 doesn't feel very user friendly. I'd expect most people would only use a consistent subset so downloading everything is slow and wasteful. 
* 2 feels the most intuitive with the current API and flexible. Seconding @Vaibhavs10's questions about finetuning, pushing to the hub and loading finetuned weights. If we load model weights from `mms-1b-fl102` and want our own finetuned adapter weights, how do I specify when loading and how is this information saved? How would we differentiate weights such that when I call `model.push_to_hub` the adapter weights are uploaded separately from the rest of the model (pattern matching?) Should the adapter weights be tied to a specific version of the 'base model' weights?
* 3 Probably simplest to do - but seems like a waste with many repeated weights.

**Comment by patrickvonplaten** [comment]
I'll leave more in-detail functionality for fine-tuning adapter weights for a future PR, but in short we can already do the following:

```py
from transformers import Wav2Vec2ForCTC

ckpt = "patrickvonplaten/mms-1b"
model = Wav2Vec2ForCTC.from_pretrained(ckpt, num_attn_adapters=1, vocab_size=277)

adapter_keys = set(model._adapters.keys())
for name, param in model.named_parameters():
    if name not in adapter_keys:
        param.requires_grad = False
```

So once we add adapter fine-tuning to the wav2vec2 fine-tuning script, we could also add a simple "freeze_all_but_adapter()" function or something.

File: src/transformers/models/wav2vec2/configuration_wav2vec2.py
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/configuration_wav2vec2.py:185]:
```suggestion
            adapters is [facebook/mms-1b-all](https://huggingface.co/facebook/mms-1b-all).
```

File: src/transformers/models/wav2vec2/tokenization_wav2vec2.py
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/tokenization_wav2vec2.py:153]:
```suggestion
            nested vocabulary such as [facebook/mms-1b-all](https://huggingface.co/facebook/mms-1b-all).
```

**Comment by patrickvonplaten** [comment]
The code is now finished. I still need to upload the adapters for the smaller checkpoints, transfer them to Facebook and write some nice docs.

File: docs/source/en/model_doc/mms.mdx
Comment (patrickvonplaten) [inline docs/source/en/model_doc/mms.mdx:38]:
```suggestion
- MMS can load different language adapter weights for different languages via [~`Wav2Vec2PreTrainedModel.load_adapter`]
```

File: tests/models/wav2vec2/test_tokenization_wav2vec2.py
Comment (patrickvonplaten) [inline tests/models/wav2vec2/test_tokenization_wav2vec2.py:776]:
```suggestion
    @require_torchaudio
    def test_nested_vocab(self):
```

File: tests/models/wav2vec2/test_tokenization_wav2vec2.py
Comment (patrickvonplaten) [inline tests/models/wav2vec2/test_tokenization_wav2vec2.py:776]:
```suggestion
```

File: tests/models/wav2vec2/test_modeling_wav2vec2.py
Comment (patrickvonplaten) [inline tests/models/wav2vec2/test_modeling_wav2vec2.py:1868]:
```suggestion
    @require_torchaudio
    def test_inference_mms_1b_all(self):
```

File: docs/source/en/model_doc/mms.mdx
Comment (patrickvonplaten) [inline docs/source/en/model_doc/mms.mdx:48]: Will fill out once model ids are moved to the Facebook org

File: src/transformers/models/hubert/modeling_hubert.py
Comment (patrickvonplaten) [inline src/transformers/models/hubert/modeling_hubert.py:608]: Note that the changes here and in all other modeling files that are != modeling_wav2vec2.py are just added so that the # Copied from logic can stay in place

**Comment by patrickvonplaten** [comment]
Could I get a final review here @sgugger @amyeroberts ? Once approved, I'll move the facebook checkpoints, add some more examples to the docs & fix the doc test.

The final question here for me is whether I should:
a) Keep # Copied from at the expense of adding currently not used code to Hubert etc...
b) Remove # Copied fromr
c) Adapt the config of Hubert etc.. as well so that one could fine-tune Hubert with this adapter training going forward.

Currently I have a) implemented as it's the safest option IMO. Happy to hear your opinion though.

**Comment by LYPinASR** [comment]
'Wav2Vec2Processor' object has no attribute 'set_lang'

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py
Comment (sanchit-gandhi) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:1065]: These weights + biases can then be regular `nn.Linear` layers

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py
Comment (sanchit-gandhi) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:1063]: Are we good hard-coding this? Or better to set it based on the dimensionality of the `hidden_states` in case the model dims are changed?

**Review by sanchit-gandhi — APPROVED** [review]
Loading function is clean 👌 Few small nits regarding the adapter layer, but overall like the design decisions taken:
1. Good for me to keep the `# Copied from...` statements across the CTC models, but will leave this to a core maintainer to give their views
2. Agree with the adapter weight loading API

**Review by sgugger — COMMENTED** [review]
LGTM! Would be nice if the `load_adapter` could be refactored in some central module since it has a lot of logic in common with `from_pretrained`, but this can be done if we start having more of those later down the road.

**Review by sgugger — APPROVED** [review]
LGTM! Would be nice if the `load_adapter` could be refactored in some central module since it has a lot of logic in common with `from_pretrained`, but this can be done if we start having more of those later down the road.

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py
Comment (sanchit-gandhi) [reply src/transformers/models/wav2vec2/modeling_wav2vec2.py:1063]: Yep agree! Was also thinking a config param would be good here (one that's tied to the `hidden_states` dim)

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py
Comment (patrickvonplaten) [reply src/transformers/models/wav2vec2/modeling_wav2vec2.py:1075]: Yes

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py
Comment (patrickvonplaten) [reply src/transformers/models/wav2vec2/modeling_wav2vec2.py:1063]: Moved it to the config

**Review by amyeroberts — APPROVED** [review]
Really nice PR and new model abilities. Thanks for adding! Tests especially were clear and helped show the expected behaviour 🤗 

Mostly just nits - main comment is some missing asserts in the tests. 

I'm not sure sure about the added logic and layers to the models copying from wav2vec2. Classic inheritance problem but it's a bit counter intuitive to have calls to a method -- `load_adapter` -- which the model doesn't have in the modeling code. Not a big issue -- if we find it confuses users, then we can handle e.g. `def load_adapter` which raises `NotImplementedError`. Noting just as a limitation of adapting models with copying logic.

File: tests/models/wav2vec2/test_modeling_wav2vec2.py
Comment (patrickvonplaten) [reply tests/models/wav2vec2/test_modeling_wav2vec2.py:300]: Hmm logging is arguably better with just `assert` actually

Comment (patrickvonplaten) [reply tests/models/wav2vec2/test_modeling_wav2vec2.py:300]: But not a big deal, let's change it

Comment (amyeroberts) [reply tests/models/wav2vec2/test_modeling_wav2vec2.py]: OK, I don't have a strong opinion, so if you have a preference I'm happy to go with that

File: src/transformers/models/wav2vec2/convert_wav2vec2_original_pytorch_checkpoint_to_pytorch.py
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/convert_wav2vec2_original_pytorch_checkpoint_to_pytorch.py:168]:
```suggestion
            return is_used
```

Comment (patrickvonplaten) [reply src/transformers/models/wav2vec2/convert_wav2vec2_original_pytorch_checkpoint_to_pytorch.py]: True correct it, good catch!

**Pull request merged**
